llvm/lib/Transforms/Utils/ModuleUtils.cpp

```diff
@@ -208,10 +209,34 @@ void llvm::setKCFIType(Module &M, Function &F, StringRef MangledType) {
   std::string Type = MangledType.str();
   if (M.getModuleFlag("cfi-normalize-integers"))
     Type += ".normalized";
+
+  uint32_t OutHash = static_cast<uint32_t>(llvm::xxHash64(Type));
+  auto T = Triple(Twine(M.getTargetTriple()));
+  if (T.isX86() && T.isArch64Bit() && T.isOSLinux()) {
+    // Estimate the function's arity (i.e., the number of arguments) at the ABI
+    // level by counting the number of parameters that are likely to be passed
+    // as registers, such as pointers and 64-bit (or smaller) integers. The
+    // Linux x86-64 ABI allows up to 6 parameters to be passed in GPRs.
+    // Additional parameters or parameters larger than 64 bits may be passed on
+    // the stack, in which case the arity is denoted as 7.
+    size_t NumParams = F.arg_size();
+    bool MayHaveStackArgs = NumParams > 6;
+
+    for (unsigned int i = 0; !MayHaveStackArgs && i < NumParams; ++i) {
+      const llvm::Type *PT = F.getArg(i)->getType();
+      if (!(PT->isPointerTy() || PT->getIntegerBitWidth() <= 64))
```

Comment:
Is this condition equivalent to what I wrote in `CodeGenModule::CreateKCFITypeId` with `typeof(*PT) = clang::Type`? Specifically:
```
      if (!(PT->isPointerType() || (PT->isIntegralOrEnumerationType() &&
                                    getContext().getTypeSize(PT) <= 64)))
```

Reply:
Front end like Clang has solved it already. I think we can simply checking the number.

Reply:
It appears that clang does not reserve stack for large arguments and instead this is done later by the LLVM X86 backend. For example:
```C
struct S {
    int *p1;
    int *p2;
    int array[8];
};

int foo(struct S s, struct S *sp) {
    return *s.p1 + *s.p2 + *sp->p1 + *sp->p2;
}
```
Then when I compile to LLVM IR I see:
```
define dso_local i32 @foo(ptr noundef byval(%struct.S) align 8 %s, ptr noundef %sp) #0 {
```
Which suggests an arity of 2. But the X86 backend transforms `foo` to pass `s` on the stack, and then `sp` becomes the sole argument and is passed in `rdi`. Hence, by the chart in the PR description, this should be treated as an arity-7 function:

| Arity Indicator | Description |
| --------------- | --------------- |
| 0 | 0 parameters |
| 1 | 1 parameter in RDI |
| 2 | 2 parameters in RDI and RSI |
| 3 | 3 parameters in RDI, RSI, and RDX |
| 4 | 4 parameters in RDI, RSI, RDX, and RCX |
| 5 | 5 parameters in RDI, RSI, RDX, RCX, and R8 |
| 6 | 6 parameters in RDI, RSI, RDX, RCX, R8, and R9 |
| 7 | At least one parameter may be passed on the stack |

This predicate:
```
      // typeof(*PT) = llvm::Type
      if (!(PT->isPointerTy() || PT->getIntegerBitWidth() <= 64))
        MayHaveStackArgs = true;
```
should prevent `s` from being counted as a register argument and correctly set the arity field to 7.

Reply:
Right, `byval` is an exception. You can use `hasPassPointeeByValueCopyAttr()` to check it.

Reply:
Thank you for the suggestion. I looked at `llvm::Argument::hasPassPointeeByValueCopyAttr()`, but it looks like it is only available where a function is being defined. It does not appear to be available where a call is made through a function pointer. Therefore, I'm not sure that `llvm::Argument::hasPassPointeeByValueCopyAttr()` will be helpful since KCFI requires the ID to be computed identically at both the call site and the call target.

Or, do you think I am overlooking something, and that there is a way to use `llvm::Argument::hasPassPointeeByValueCopyAttr()` or something similar at an indirect call site? As far as I can tell, the only information that is available at an indirect call site is the function pointer type, which does contain the number of arguments and their types, but does not appear to contain an indication as to whether an argument may be passed on the stack.

Reply:
It is an argument attribute, should be identical when defined and called, otherwise, we will have mismatch issue. I assume a simply use [like](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Utils/InlineFunction.cpp#L1677) this should be ok.

Reply:
Hi @phoebewang, KCFI only computes hashes for indirect calls, not direct ones. The example you cited uses `CallBase::getCalledFunction()`, whose documentation reads "Returns the function called, or null if this is an indirect function invocation or the function signature does not match the call signature."

Reply:
Sorry, I don't understand the point of indirect calls. Here `setKCFIType` has an argument `F` which never will be null. Why do we need another `getCalledFunction()`?